crates/db/src/traits.rs

```diff
@@ -194,39 +193,43 @@ pub enum BlockStatus {
     Invalid,
 }
 
-/// Db trait for the (consensus layer) chain state database.  For now we only
-/// have a modestly sized "toplevel" chain state and no "large" state like the
-/// EL does.  This trait is designed to permit a change to storing larger state
-/// like that in the future without *too* much extra effort.  We decide new
-/// states by providing the database with a generic "write batch" and offloading
-/// the effort of deciding how to compute that write batch to the database impl.
+/// Low-level Strata chainstate database.  This provides the basic interface for
+/// storing and fetching write batches and toplevel states on disk.
+///
+/// Currently we do not have a "bulk" state that we would want to avoid storing
+/// in memory all at once.  In the future, we expect that this interface would
+/// be extended to expose a "finalized" state that's fully materialized, along
+/// with functions to walk the finalized state forwards and backwards.  We can
+/// use the unmerged write batches to construct a view of more recent states
+/// than the fully materialized state in-memory.
+///
+/// For now, the full state is just the "toplevel" state that can always be
+/// expected to be of moderate size in memory.
+// TODO maybe rewrite this around storing write batches according to blkid?
 pub trait ChainstateDatabase {
     /// Writes the genesis chainstate at index 0.
-    fn write_genesis_state(&self, toplevel: &Chainstate) -> DbResult<()>;
+    fn write_genesis_state(&self, toplevel: Chainstate) -> DbResult<()>;
```

Comment:
Won't passing the `Chainstate` instead of `&Chainstate` here incur an extra `clone()` on the caller side?

Reply:
In theory yes but because of how the ops wrapper works this isn't ideal, ultimately because you can't pass a ref between `Future`s.  There should be a ticket to tweak the ops interface to *allow* you to pass by-ref in blocking calls if you want to where it's possible, but take by-value in async calls.